packages/dataviews/src/components/dataform/stories/index.story.tsx

```diff
@@ -118,3 +124,44 @@ export const Default = ( { type }: { type: 'panel' | 'regular' } ) => {
 		/>
 	);
 };
+
+export const CombinedFields = ( { type }: { type: 'panel' | 'regular' } ) => {
+	const [ post, setPost ] = useState( {
+		title: 'Hello, World!',
+		order: 2,
+		author: 1,
+		status: 'draft',
+	} );
+
+	const form = {
+		fields: [ 'title', 'status_and_visibility', 'order', 'author' ],
+		layout: {
+			combinedFields: [
+				{
+					id: 'status_and_visibility',
+					label: 'Status & Visibility',
+					children: [ 'status', 'password' ],
+					direction: 'vertical',
+					render: ( { item } ) => item.status,
```

Comment:
What is this function about?

The thing that I'm wondering personally is whether for "forms" these are combined fields or more things like "groups"...

Reply:
I think having more example of how this translates in actual forms and panels would help make the right abstraction.

Reply:
> What is this function about?

Its suppose to be the same as the `render` function of the field config, but its acting more like the `getValue` function.

> The thing that I'm wondering personally is whether for "forms" these are combined fields or more things like "groups"...
> I think having more example of how this translates in actual forms and panels would help make the right abstraction.

Groups would probably fit this use case better, the main example is the status field that includes the status radio control, posts schedule, the password field and also the sticky checkbox: https://github.com/WordPress/gutenberg/blob/2943dd3adfa4e1ad33b980760277377df9ac0f91/packages/editor/src/components/post-status/index.js#L197

https://github.com/user-attachments/assets/b7d2caf1-719e-4c2d-9bc0-46ea4647fce2

We could also create a single status field that includes all of this logic, we did have to remove the `radio` dependency here: https://github.com/WordPress/gutenberg/blob/3ec1ced865757f6568c42c126acb3478b25fe2be/packages/edit-site/src/components/post-fields/index.js#L283-L294

This is mostly a use case for the panel view. For the regular form one could render multiple regular data forms to create the grouping. Although there would be some additional use cases like the column use case for displaying two fields beside each other ( we have a similar use case in the product form for regular and sale price ):
![Screenshot 2024-09-18 at 10 35 38 AM](https://github.com/user-attachments/assets/3ec89040-14be-4ab3-a1e0-6c7b093942a0)

Reply:
> The thing that I'm wondering personally is whether for "forms" these are combined fields or more things like "groups"...

Would you envision "groups" to be like field types that accept children, or act in a similar manner as the combinedField API?

Reply:
> Its suppose to be the same as the render function of the field config, but its acting more like the getValue function.

Do you think we can guess this from the "fields" or is the "render" function mandatory here? I'm also curious how the "combined fields" or "groups" feature translate in regular forms. What's the output there?

Reply:
> Do you think we can guess this from the "fields" or is the "render" function mandatory here?

In this specific example it only shows the `status` value in the panel view so the `render` or a `getValue` would be mandatory.
If we wanted, we could guess it from the fields and show a comma delimited list of the nested field values.
The panel view is an exception here as it separates the edit view from the visual.

> I'm also curious how the "combined fields" or "groups" feature translate in regular forms. What's the output there?

In regular forms the main difference would be a title for the grouped fields and the option to display them in a column layout. Similar to the JSON Forms layout examples: https://jsonforms.io/examples/layouts/#group

Reply:
> In regular forms the main difference would be a title for the grouped fields and the option to display them in a column layout. Similar to the JSON Forms layout examples: https://jsonforms.io/examples/layouts/#group

I think that makes sense. Is it already working in this PR?  I think my main concern here is to avoid special cases for layouts (panel or default)

Reply:
I didn't update the `default` data form yet, let me update it to support that.

Reply:
I updated it to support the `regular` data form type as well as part of this commit: https://github.com/WordPress/gutenberg/pull/65399/commits/42131f339ce26aab62a8ec5b06c7179e093bccd9

Reply:
Is this PR working for the `regular` form? I don't see any visual grouping there:

<img width="1044" alt="Captura de ecrã 2024-09-25, às 17 16 45" src="https://github.com/user-attachments/assets/1d612f01-e839-4ae8-a722-d435eb87e7a0">

Does it make sense to render these as a `fieldset` being the combined field's label its `legend`?

Reply:
Based on what the problem we want to solve, there's two related questions:

- being able to stack fields horizontally or vertically, even if they are not "combined fields"
- being able to have to provide a different field combination depending on a field value:
  - combine status+password if status' value is `draft`
  - combine status+date if status' value is `publish`
  - only show status if value is `private`

How would those work with the combinedFields API introduced here?

Reply:
> Is this PR working for the regular form? I don't see any visual grouping there:

Yeah that would require extra styling, the only visible option in the `regular` form is the addition of the `Status & Visibility` label underneath the title. But there is no clear distinction of when that ends.

> Does it make sense to render these as a fieldset being the combined field's label its legend?

Yeah `fieldset` makes sense I will update that.

Reply:
> How would those work with the combinedFields API introduced here?

Good questions 👍 

> being able to stack fields horizontally or vertically, even if they are not "combined fields"

Currently I followed the DataViews and support the `direction` prop that will render it horizontally or vertically. Currently its hard to do if they are not "combined fields" we did have to merge them into a single field or nest DataForms.
It may be worth pulling this out of combined fields and have a specific horizontal and vertical layout, similar to what JSON Forms does: https://jsonforms.io/examples/layouts/#horizontal-layout ?

One question I did have is for the direction, does that always look the same in the `regular` form and the `panel` form, or may a form overwrite this? This may be a ["moo" point](https://www.reddit.com/r/howyoudoin/comments/r79qck/its_a_moo_point/) though as storybook may be the only place where the form type gets toggled on a constant basis.

> being able to have to provide a different field combination depending on a field value:

I currently did that by updating the form object directly and updating the `children` if in this case the status criteria changes: https://github.com/louwie17/gutenberg/pull/1/files#diff-4ed486e973856f952d2b80aea381fc1f5a1d0004995a0b68554147b02cee0b55R76-R87

Now this isn't so bad when talking about a very small form as it is now, but I don't like this idea if we have a lot more conditional logic that depends on different data ( meaning the form object gets regenerated a lot ).

I was kind of thinking that just like the `isValid` function we may want a `isVisible` function, that in the future will also support the JSON Schema validation rules in support of the[ rule parser](https://github.com/WordPress/gutenberg/pull/65171) that Nadir is working on. This way visibility can be handled on a per field level and avoids re-rendering of the entire form.

Reply:
> One question I did have is for the direction, does that always look the same in the regular form and the panel form, or may a form overwrite this?

My approach is to think about the API we want to introduce at a conceptual level. In this case, we're trying to introduce the idea of "grouping fields". Does that feature make sense for both layouts? It think it does, even if it means different things for different layouts at a visual level. There may be cases where an API/feature only makes sense for a given layout. It'd be ok to ignore the other layout in that case. Does this help?

> I was kind of thinking that just like the isValid function we may want a isVisible function

I like this direction better than the current one.

Perhaps we can start from this angle? This could be a separate PR to implement the visibility feature. This PR's scope could be just about clarifying the API to group fields. (both things can land & be worked in parallel).

Reply:
Agree about decoupling visibility/conditional rendering and grouping.

I will add that ultimately, I see different level or ways to group fields: tabs, vertical or horizontal fieldsets (kind of like here) and maybe other ways we're not seeing right now. Just food for thought to come up with the right approach. 

Something to consider as well is that even the `layout.type` can be seen as "not necessary" and instead we could just imagine something like:

```
const fields = [ 
   { field: 'author', control: 'panel-row' }, 
   { field: 'title', control: 'inline' }, 
   { field: 'some group', control: 'group', fields: [ 'child1', 'child2' ] }
```

in other words, instead of having a layout type, you can just mix and match how to render the fields. In the same form you can have some inline controls, grouped or not and other fields that behave like the panel (dropdown)...

Reply:
> in other words, instead of having a layout type, you can just mix and match how to render the fields. In the same form you can have some inline controls, grouped or not and other fields that behave like the panel (dropdown)...

Hmm yeah that could work, we did have to restructure our current panel and regular view to handle a single field instead of a list of fields and render each layout type respectively of the field.
With this principle the `group` structure makes a bit more sense, but I wonder if the use case of combining panel and regular/inline fields would be heavily used.
Makes me still think we could use `layout.type` as the default.